File: 26_dp_ii/01_min_path_sum.py
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        return self.helper(grid, 0, 0, {})

    def helper(self, grid, r, c, dp):
        if (r, c) in dp:
            return dp[(r, c)]

        if r == len(grid) - 1 and c == len(grid[0]) - 1:
            return grid[r][c]

        if r == len(grid) or c == len(grid[0]):
            return float("inf")

        p1 = self.helper(grid, r + 1, c, dp)
        p2 = self.helper(grid, r, c + 1, dp)

        dp[(r, c)] = grid[r][c] + min(p1, p2)
        return dp[(r, c)]

File: 26_dp_ii/01_min_path_sum.rs
use std::{cmp::min, collections::HashMap};

impl Solution {
    pub fn min_path_sum(grid: Vec<Vec<i32>>) -> i32 {
        let mut dp: HashMap<(usize, usize), i32> = HashMap::new();
        return Solution::helper(&grid, 0, 0, &mut dp);
    }

    fn helper(
        grid: &Vec<Vec<i32>>,
        r: usize,
        c: usize,
        dp: &mut HashMap<(usize, usize), i32>,
    ) -> i32 {
        if dp.contains_key(&(r, c)) {
            return *dp.get(&(r, c)).unwrap();
        }
        if r == grid.len() - 1 && c == grid[0].len() - 1 {
            return grid[r][c];
        }

        if r == grid.len() || c == grid[0].len() {
            return i32::MAX;
        }

        let p1 = Solution::helper(grid, r + 1, c, dp);
        let p2 = Solution::helper(grid, r, c + 1, dp);
        let res = grid[r][c] + min(p1, p2);

        dp.insert((r, c), res);
        return res;
    }
}
